Support layer compositing and alpha, mk2

rsheeter opened the pull request:
Take two (take 1 was #35) at layer compositing and alpha. Changes layer structure to a DAG based on chat with @behdad and @drott. 

- Modes are as per https://www.w3.org/TR/compositing-1/
   - Most implementations already support most modes
      - D2D is notable in that it seems to have relatively few of the modes
      - sources
         - https://developer.apple.com/documentation/coregraphics/cgblendmode?language=objc
         - https://www.cairographics.org/operators/
         - https://docs.microsoft.com/en-us/windows/win32/api/d2d1_1/ne-d2d1_1-d2d1_composite_mode
         - https://source.chromium.org/chromium/chromium/src/+/master:third_party/skia/include/core/SkBlendMode.h

Fixes #13.

The diff is a bit hard to read, https://github.com/googlefonts/colr-gradients-spec/blob/rs/colr-gradients-spec.md#compositing shows the updated md.

Edit: I believe this fixes #3 given that you get layers and `PaintComposite` can be used for hole-punching via [Source Out](https://www.w3.org/TR/compositing-1/#porterduffcompositingoperators_srcout).

Edit2: Note that I changed Extend and the Paint format to `uint8` instead of `uint16`; @anthrotype pointed out that I neglected to highlight this (ty).

anthrotype commented on colr-gradients-spec.md:
```diff
@@ -337,39 +353,73 @@ struct ColorStop
+  COMPOSITE_SRC_ATOP       =  9,  // https://www.w3.org/TR/compositing-1/#porterduffcompositingoperators_srcatop
+  COMPOSITE_DEST_ATOP      = 10,  // https://www.w3.org/TR/compositing-1/#porterduffcompositingoperators_dstatop
+  COMPOSITE_XOR            = 11,  // https://www.w3.org/TR/compositing-1/#porterduffcompositingoperators_xor
+  COMPOSITE_SCREEN         = 12,  // https://www.w3.org/TR/compositing-1/#blendingscreen
```
the https://www.w3.org/TR/compositing-1/ spec lists "screen" under the blend modes rather than a porter-duff compositing operators. The way you split the two sections implies it belongs to the latter.

rsheeter replied:
Fixed, ty.

anthrotype commented on colr-gradients-spec.md:
```diff
@@ -337,39 +353,73 @@ struct ColorStop
 };
 
-enum Extend : uint16
+enum Extend : uint8
```
i'm ok to save 1 byte here and in the paint formats (why where they uint16 to begin with?).
Maybe you could make changes like these to pre-existing types more prominent in the main PR message, as that will help implementers (me) updating existing draft implementations.

rsheeter replied:
Will do, good catch.

anthrotype commented on colr-gradients-spec.md:
```diff
@@ -392,23 +442,42 @@ struct PaintRadialGradient
-typedef ArrayOf<LayerV1Record, uint32> LayerV1List;
+// Clip paint to the region inked by gid
+// Paint must be unbounded
+struct PaintGlyph
```
what is the use of `PaintGlyph`? I don't see it presented or discussed anywhere.

rsheeter replied:
It basically replaces LayerV1Record: paint a non-color glyph using the specified paint. I will update the comment to hopefully be more informative.

anthrotype replied:
Ah, I didn't notice you also removed the LayerV1Record.
I am a bit confused by the characterization of PaintGlyph and PaintColrGlyph structures as "paints" instead of just "layers". In earlier drafts and so far, we were thinking of a layer as a structure connecting a glyph (via its gid) and some paint (a color, or gradient). Now the distinction between layer and paint seems to be blurred.
Early on you say that base glyphs are mapped to "list of layers", however the LayerV1List is now an array of offsets to Paint structures, not layers strictly speaking. 

One question. What happens if these offsets in LayerV1List is directly one of the old format 1, 2 and 3 "pure" or "unbounded" Paints that don't have a glyph index associated with them? Should we disallow these in the top level and require they be children of PaintGlyph?

rsheeter replied:
I've been deliberately light on restrictions. I suspect we do need to add some, but hopefully biasing towards having as few as possible. 

I don't think you want to require the unbounded paints be children of PaintGlyph; PaintComposite {PaintGlyph, mode, Paint Solid or Gradient} should be valid. I'm not completely clear if the extent to fill can be given a clear definition when an unbounded paint is never bounded/clipped.

@behdad @drott wdyt?

rsheeter replied:
Behdad responded below, and I believe we've come to agreement on boundedness; resolving. As always, please reopen if not adequately addressed.

anthrotype commented on colr-gradients-spec.md:
```diff
@@ -252,27 +254,41 @@ inverse-transform approach noted above can fall back to a linear-gradient
-A group allows a set of layers to be composited independently and then have the
-final result composited into the end result as a single layer with alpha. This
-is similar to a `group` with `opacity` in an SVG.
+The alpha channel for a layer can be populated using `PaintComposite`:
```
I don't see where this "alpha channel for a layer" is defined.. I suspect I'm missing something basic about these compositing and blending modes.

rsheeter replied:
Via the A part of the colors used. [Source in](https://www.w3.org/TR/compositing-1/#porterduffcompositingoperators_srcin) defines color exclusively in terms of the source color while combining alpha so you can do things like combine my source with a solid paint whose alpha is <value> or do fancy things like combine with a gradient whose stops have different alphas.

anthrotype replied:
ah, so you mean for PaintSolid you set it from the A of the RGBA color in the CPAL table. That means it won't be variable, unlike the alpha value for the gradient paints.

anthrotype reviewed (commented):
I left a few comments below.

Incidentally, the "High-level Design" section at the top needs updating as it currently says "There are four types of paint defined currently..." whereas this PR defines more.

anthrotype commented on colr-gradients-spec.md:
```diff
@@ -252,27 +254,41 @@ inverse-transform approach noted above can fall back to a linear-gradient
-# Group
+# Alpha
 
-A group allows a set of layers to be composited independently and then have the
```
without the PaintGroup how do we solve the problem of multi-layer opacity discussed #32?

rsheeter replied:
You would use `PaintComposite` to combine `PaintColrGlyph` with a `Paint` that populates alpha.

rsheeter commented:
Updated high level design, ty @anthrotype for noticing I hadn't yet.

anthrotype commented on colr-gradients-spec.md:
```diff
@@ -76,10 +76,18 @@ COLR table is extended to expose a new vector of layers per glyph.  If a glyph
-layer includes a glyph id of shape as well as a *paint*.  There are four types
-of paint defined currently, with capacity for future extensions: solid, linear
-gradient, radial gradient, and group.
+A glyph using the new extension is mapped to a list of layers. Each layers is formed by a directed acyclic graph of paints. Several different types of paint
```
```suggestion
A glyph using the new extension is mapped to a list of layers. Each layer is formed by a directed acyclic graph of paints. Several different types of paint
```

rsheeter replied:
Fixed

behdad commented:
The part saying:
```
Glyph draws a non-COLR glyph, filled using a type 1, 2, or 3 paint.
```
I don't think the limitation is necessary.

behdad commented:
Looks good to me. I think the part about bounded vs unbounded paints needs more work.

To answer Cosimo's comment: in the new design we have one type of thing, called Paint. It's the same as layer, or image. Any paint is a full image. It is either bounded or unbounded. The following needs incorporation into the spec:

- Each layer in a `LayerV1List` must be bounded.

- Solid and gradient paints are unbounded.

- A PaintTransformed is bounded if and only if the source is bounded,

- PaintGlyph and PaintColrGlyph results are always bounded. The source for PaintGlyph does *not* have to be unbounded. So that needs fixing in the current PR.

- PaintComposite result's boundedness is a bit more complicated as it's a function of the boundedness of src & backdrop as well as the CompositeMode. This is for example addressed similarly in the cairo documentation, although only in relationship to the source (if I recall correctly): https://www.cairographics.org/operators/

I try to enumerate what's needed to be speced here (please double-check my work):

The following CompositeModes are always bounded: CLEAR.
The following CompositeModes are bounded if and only if src is bounded: SRC, SRC_OVER, SRC_OUT.
The following CompositeModes are bounded if and only if backdrop is bounded: DEST, DEST_OVER, DEST_OUT.
The following CompositeModes are bounded if and only if *either* src or backdrop are bounded: SRC_IN, DEST_IN.
All the other CompositeModes are bounded if and only if *both* src or backdrop are bounded.

rsheeter commented:
@behdad updated per feedback, PTAL. My list of boundedness differs only for SRC/DEST_OVER, which I have as bounded if src *and* backdrop are bounded.

While I'm at it, summon @anthrotype @drott!

drott commented:
I did a longer first pass today and here's some comments, first the important ones which I think we need to address in this CL first, then some considerations we may want to do in subsequent edits. At the bottom, I'll be pasting a sketch of the algorithm I have in mind for drawing this.

#### For this CL:
1. I think we need a pre-computed bounding box. After discussing with Cosimo we came to the following idea: Mac COLR v0 implementation uses two points at the extremes in the `glyf` table to size the allocation. Cosimo and I considered doing the the same here. Background: For COLR v0 and COLR v1 so far without transforms and blend modes, the union of layers suffices, and was rather fast to compute. For v1 with the edits in this CL, the root layer extents are not easy to compute any more. We need to describe what should be the top level surface extent, and how to retrieve it from the gid entry of the `glyf` table and use a rectangle that fits the extreme. Ideally that rectangle should be tight around the color glyph shape, as it controls allocation size.
2. There is a mismatch between `PaintComposite`'s terminology: **src** and **backdrop** and the blending modes terminology: **src** and **destination**. We should make that consistent to use only one or the other.

#### In subsequent edits:

1. Offsets of various `PaintXYZ`: Do these offsets still refer to the beginning of the parent struct, or do we want all of them to point to a common reference point for one COLR v1 glyph? If it's a graph, to me it feels slightly confusing to keep track of the parents, but might be subjective - what do others think?
2. Reuse of individual `PaintGlyph`, `PaintTransformed`, `PaintComposite` etc: Can one COLR v1 glyph refer to PaintGlyph "objects" that come from a different COLR v1 glyph? Or is our "maximum" unit of reuse `PaintColrGlyph`?
3. In Chrome we support glyph intercepts for ink skipping in the underline: Can these still be efficiently computed with path operations, or only from looking at which pixels are transparent on the root layer?
4. So far `PaintGlyph` aways has _ClipIn_ semantics, i.e. fill the inside of the shape with what's referenced as Paints from PaintGlyph: Do we need _ClipOut_ semantics? - Current thinking: No, probably not, as you could achieve the same by subtracting the desired shape from the extent box.
5. We should rephrase "High Level Design" to also talk about how we model that to available primitives in a 2D graphics library, then perhaps regroup Transforms and Composites under the heading "Graphical Primitives" and potentially rename the section. I can do that in subsequent edits.

#### Sketch of algorithm (perhaps include this in a prose section of the spec?), feedback or corrections welcome:

```
Allocate a bitmap for the glyph according to glyf table entry extends for gid
0) Traverse layers for gid, add current gid to blacklist *1
 a) Paint a paint, switch:
    1) PaintGlyph
         gid must not COLRv1
         saveLayer
         setClipPath to gid path
           recurse to a)
         restore
    2) PaintColorGlyph
         gid must be from 
         if gid on recursion blacklist, do nothing
         recurse to 0) with different gid
    3) PaintComposite
          paint Paint for backdrop, call a)
          saveLayer() with setting composite mode, on SkPaint
          paint Paint for src, call a)
          restore with save composite mode
    4) PaintTransformed
          saveLayer()
          apply transform 
          call a) for paint
          restore
    5) PaintRadialGradient
          SkCanvas::drawPaint with radial gradient configured
          (expected to be bounded by parent composite mode or clipped by current clip, check bounds?)
    6) PaintLinearGradient
          SkCanvas::drawPaint with liner gradient configured
          (expected to be bounded by parent composite mode or clipped by current clip, check bounds?)
    7) PaintSolid
          SkCanvas::drawColor with color configured

 *1 (in the implementaiton, potentially collapse each pair as PaintComposite (3) with first gid as backdrop, second as src, or just draw on root SkCanvas directly)    
```

behdad commented:
> My list of boundedness differs only for SRC/DEST_OVER, which I have as bounded if src _and_ backdrop are bounded.

Correct. Thanks.

rsheeter commented:
Ty @drott! I have some edits to make (once meeting hell ends); a few immediate comments:

> 1. I think we need a pre-computed bounding box

I like this idea, will update.

> There is a mismatch between PaintComposite's terminology: src and backdrop and the blending modes terminology: src and destination.

https://www.w3.org/TR/compositing-1 refers to backdrop. If this is OK to cite from the font spec then I suggest we use the term backdrop. Does that seem reasonable?

> 2. Reuse of individual `PaintGlyph`, `PaintTransformed`, `PaintComposite` etc: Can one COLR v1 glyph refer to PaintGlyph "objects" that come from a different COLR v1 glyph? Or is our "maximum" unit of reuse `PaintColrGlyph`?

Reuse any paint the font compiler wants. I think of the DAG as picking parts from a common pool.

behdad commented:
@rsheeter please add yourself to the authors as well. It feels weird every time opening your draft and seeing my name and Dominik's there only.

behdad commented:
Thanks Dominik. Replying to some of the easier ones:

> 1. Offsets of various `PaintXYZ`: Do these offsets still refer to the beginning of the parent struct, or do we want all of them to point to a common reference point for one COLR v1 glyph? If it's a graph, to me it feels slightly confusing to keep track of the parents, but might be subjective - what do others think?

The offsets are from the beginning of the current paint.

> 2. Reuse of individual `PaintGlyph`, `PaintTransformed`, `PaintComposite` etc: Can one COLR v1 glyph refer to PaintGlyph "objects" that come from a different COLR v1 glyph? Or is our "maximum" unit of reuse `PaintColrGlyph`?

Yes they can, by way of byte reuse. That's the beauty of this approach: the builder doesn't need to even identify duplicate paints. The compiler reuses bytes that are similar.

> 3. In Chrome we support glyph intercepts for ink skipping in the underline: Can these still be efficiently computed with path operations, or only from looking at which pixels are transparent on the root layer?

It can *if* you define what the shape of a gradient is. Or assume gradients to be opaque and then is not much different from what you currently do; just additional transform...  But yeah in this case pixel check may be easier.

> 4. So far `PaintGlyph` aways has _ClipIn_ semantics, i.e. fill the inside of the shape with what's referenced as Paints from PaintGlyph: Do we need _ClipOut_ semantics? - Current thinking: No, probably not, as you could achieve the same by subtracting the desired shape from the extent box.

Correct.

> 5. We should rephrase "High Level Design" to also talk about how we model that to available primitives in a 2D graphics library, then perhaps regroup Transforms and Composites under the heading "Graphical Primitives" and potentially rename the section. I can do that in subsequent edits.

Sgtm.

> 6. Do we want to keep the `LayerV1List` or simplify this to using `PaintComposite` at the root already? Feels slightly redundant but may be useful as a storage optimisation.

What Rod said. The list of layers is also more efficient since new layer can be painted right on top of the existing output. Same might be doable with the DAG but needs more analysis to implement that way.

rsheeter commented:
Added a note about base glyf bbox, a note about offsets being relative to the containing struct, and captured @drott rendering pseudocode.

I believe I have addressed all of @drott notes except the one that includes the explicit offer to make the changes :)

> 5. We should rephrase "High Level Design" to also talk about how we model that to available primitives in a 2D graphics library, then perhaps regroup Transforms and Composites under the heading "Graphical Primitives" and potentially rename the section. I can do that in subsequent edits.

behdad commented:
> 1. I think we need a pre-computed bounding box. After discussing with Cosimo we came to the following idea: Mac COLR v0 implementation uses two points at the extremes in the `glyf` table to size the allocation. Cosimo and I considered doing the the same here. Background: For COLR v0 and COLR v1 so far without transforms and blend modes, the union of layers suffices, and was rather fast to compute. For v1 with the edits in this CL, the root layer extents are not easy to compute any more. As graphics libraries may want to know what size to allocate in the glyph cache, we need to describe what should be the top level surface extent, and how to retrieve it from the respective gid entry of the `glyf` table and use a rectangle spanned by the extremes. Ideally that rectangle should be tight around the color glyph shape, as it controls allocation size.

That's a very good point.

Can we spec that the `glyf`/otherwise bounding box of the *first* layer is used? And then have a Note that a degenerate outline with two diagonal extrema points can be used in the first layer glyph to extend the bounding box as necessary?

behdad commented:
> Can we spec that the `glyf`/otherwise bounding box of the _first_ layer is used? And then have a Note that a degenerate outline with two diagonal extrema points can be used in the first layer glyph to extend the bounding box as necessary?

The alternative is to say that the `glyf`/otherwise bounding box of the *current* gid is used. That's what `sbix` does. Now I realize that that's probably what @drott meant. Let me read @rsheeter's change to see if that's clear or needs clarification.

behdad commented on colr-gradients-spec.md:
```diff
@@ -335,8 +339,17 @@ The following paints *may* be bounded:
+## Bounding Box
+
+To simplify implementation allocation of drawing surface the
+`BaseGlyphV1Record` `gid` entry in the `glyf` table must provide
```
This phrase: "`BaseGlyphV1Record` `gid` entry in the `glyf`" doesn't make sense and needs fixing. Removing the `BaseGlyphV1Record` might do.

drott replied:
I'd phrase that a bit more strict, perhaps:

> The surface extents on which to paint the `Paint` operations of a `LayerV1List` is defined by two on-curve points in the `glyf` table at the `gid` entry for the same `gid` as in this glyph's `BaseGlyphV1Record`: one at `(min-x, min-y)` and one at `(max-x, max-y)`.

@behdad:

> Can we spec that the glyf/otherwise bounding box of the first layer is used? And then have a Note that a degenerate outline with two diagonal extrema points can be used in the first layer glyph to extend the bounding box as necessary?

Your suggestion to spec'ing is as "bounding box of first layer": Did you mean without traversing all paints  requiring that the first paint of the first layer is bounded?

drott replied:
> The alternative is to say that the `glyf`/otherwise bounding box of the current gid is used.

Aren't we looking at `glyf` for the "current" gid? Yes, I did mean SBIX style, if I understood it correctly.

rsheeter replied:
I meant to take the bounding box for the glyph whose id is in the BaseGlyphV1Record::gid field. Is that the same as the "current" gid?

I dislike having to process the first layer; removes the benefit of a simple path to knowing the extents.

I will move the two on-curve points bit to be a note.

behdad commented on colr-gradients-spec.md:
```diff
@@ -335,8 +339,17 @@ The following paints *may* be bounded:
+
+To simplify implementation allocation of drawing surface the
+`BaseGlyphV1Record` `gid` entry in the `glyf` table must provide
+the bounding box for the final COLR glyph by providing two
```
The "by providing" should become a note, not a recipe.

behdad replied:
> not a recipe

Or, "not a requirement" I meant.

rsheeter replied:
Updated, better?

anthrotype commented on colr-gradients-spec.md:
```diff
@@ -74,10 +81,18 @@ COLR table is extended to expose a new vector of layers per glyph.  If a glyph
+1. **Solid** paints a solid color
+1. **Linear gradient** paints a linear gradient
+1. **Radial gradient** paints a radial gradient
+1. **Transformed** reuses another paint, applying an affine transformation
```
I don't remember if this has been already discussed, but what happens if PaintTransformed references one of the unbounded paints (formats 1, 2, 3)?
At least format 1 (solid paint) has no actual (x, y) coordinates to be transformed.
For linear and radial gradient paint, one could maybe apply the transformation to the gradient points.
Radial also has its own Affine2x2. Would one concatenate this with PaintTransformed?

Or was the idea that PaintTransformed be only used in combination with PaintGlyph or PaintColorGlyph?

drott replied:
I think the algorithm covers that: a `PaintTransformed` creates a layer, then applies the transformation matrix. Then child paints happen. If the child paint is a `PaintSolid`, `PaintLinearGradient`, `PaintRadialGradiant` - these would be implemented by calling Skia's `drawPaint` which fills "everything" at this stage with the fill. The gradient matrices (for radial) are set up with the point info taken as is from the struct. (the points inverted with that matrix). So effectively, yes, the transformations of outer `PaintTransformed` and inner gradient matrices are combined. 

Is that the right understanding, @behdad?

anthrotype replied:
further below, in "Bounded Layers Only" section, where it says

> Every entry in the `LayerV1List` must define a bounded region... Unbounded layers must
not render... `PaintTransformed` is bounded IFF the source is bounded

so a PaintTransformed with an unbounded paint the top level will not render -- unless if it's in turn bounded by a parent PaintGlyph or PaintComposite -- Is that correct?

drott replied:
Yes, as discussed with Behdad: We would keep a flag on layer boundedness and collapse that upwards. If the top most layer is unbounded, don't draw.

drott commented on colr-gradients-spec.md:
```diff
@@ -392,23 +501,43 @@ struct PaintRadialGradient
 };
 
-typedef ArrayOf<LayerV1Record, uint32> LayerV1List;
+// Paint a non-COLR glyph, filled as indicated by paint.
+// Akin to the COLRv0 Layer Record.
+// Clip paint to the region inked by gid
```
Either keep
"// Paint a non-COLR glyph, ..."
and drop
"Clip paint to the region inked by gid"

Or perhaps replace with
"// Paint a non-COLR glyph by clipping to region inked by ink, filled as indicated by paint."

Nit/Minor: Not sure the "Akin to the COLRv0 Layer Record." comment helps.

rsheeter replied:
Fixed.

drott commented:
Some phrasing suggestions in previous comments, mostly looks good to me.

#### Re previous parts of the conversation:

> Terminology: destination / backdrop.

I see that in the W3C spec also both terms are used. I would like more consistency but okay for me to have `backdrop` as the struct member. Doesn't look to good to rename all operators to `..._BACKDROP_`.

> My question regarding whether to keep layer layerlist.

Makes sense, ok to keep as more compact representation than a chain of `PaintComposite` + OVER operators.

> > 1. Reuse of individual `PaintGlyph`, `PaintTransformed`, `PaintComposite` etc: Can one COLR v1 glyph refer to PaintGlyph "objects" that come from a different COLR v1 glyph? Or is our "maximum" unit of reuse `PaintColrGlyph`?
> 
> Yes they can, by way of byte reuse. That's the beauty of this approach: the builder doesn't need to even identify duplicate paints. The compiler reuses bytes that are similar.

Ok, since the offsets are from the current paint, that would mean that the compiler would have to move the reused ones to the end, so that they all can point forward, right?

rsheeter commented on colr-gradients-spec.md:
```diff
@@ -252,30 +267,90 @@ inverse-transform approach noted above can fall back to a linear-gradient
+## Bounded Layers Only
+
+Every entry in the `LayerV1List` must define a bounded region.
+Implementations must confirm this invariant. Unbounded layers must
+not render.
```
I tried that but it feels as redundant given the section starts with "Every entry in the `LayerV1List` must define a bounded region."

Alternate phrasing? OK to leave unchanged?

drott replied:
Ok to leave unchanged.

drott commented on colr-gradients-spec.md:
```diff
@@ -342,10 +343,11 @@ The following paints *may* be bounded:
-the bounding box for the final COLR glyph by providing two
-on-curve points, one at `(min-x, min-y)` and one at `(max-x, max-y)`.
+To simplify implementation allocation of a drawing surface the
+bounding box of the glyph corresponding to the `BaseGlyphV1Record`
```
I would perhaps insert "bounding box of the `glyf` table glyph" to make clear that we're looking at the other table and not mention that only in the note.

drott commented:
LGTM to merge - I will follow-up with some section re-arrangements in a separate change.
